doc/languages-frameworks/maven.md

```diff
@@ -0,0 +1,351 @@
+in stdenv.mkDerivation rec {
+  pname = "maven-demo";
+  version = "1.0";
+
+  src = builtins.fetchTarball "https://github.com/fzakaria/nixos-maven-example/archive/main.tar.gz";
+  buildInputs = [ maven ];
+
+  buildPhase = ''
+    echo "Using repository ${repository}"
+    mvn --offline -Dmaven.repo.local=${repository} package;
+  '';
+
+  installPhase = ''
+    install -Dm644 target/${pname}-${version}.jar $out/share/java
+  '';
+}
+```
+
+> We place the library in _$out/share/java_ since JDK package has a _stdenv setup hook_ that adds any JARs in the _share/java_ directories of the build inputs to the CLASSPATH environment.
+
+```bash
+❯ tree $(nix-build --no-out-link build-jar.nix)
+/nix/store/7jw3xdfagkc2vw8wrsdv68qpsnrxgvky-maven-demo-1.0
+└── share
+    └── java
+        └── maven-demo-1.0.jar
+
+2 directories, 1 file
+```
+
+## Runnable JAR
+
+You are able to run `mvn compile` but how do you get a runnable artifact without having to use something like building an UberJar ?
+
+We will use the same repository we built above (either _double invocation_ or _buildMaven_) to setup a CLASSPATH for our JAR.
+
+### CLASSPATH
+
+We will read the Maven repository and flatten it to a single list. This list will then be concatenated with the _CLASSPATH_ separator to create the full classpath.
+
+We make sure to provide this classpath to the _makeWrapper_.
+
+> readTree may be soon availble in nixpkgs via https://github.com/NixOS/nixpkgs/pull/101096
+
+```nix
+{ pkgs ? import <nixpkgs> { }, stdenv ? pkgs.stdenv, lib ? pkgs.lib
+, maven ? pkgs.maven, callPackage ? pkgs.callPackage
+, makeWrapper ? pkgs.makeWrapper, jre ? pkgs.jre }:
+let
+  readTree = dir:
+    lib.flatten (lib.mapAttrsToList (name: type:
+      if (type == "directory") then
+        readTree (dir + "/${name}")
+      else
+        dir + "/${name}") (builtins.readDir dir));
+
+  repository = callPackage ./build-maven-repository.nix { };
+  jars = lib.traceVal
+    (builtins.filter (path: (lib.hasSuffix ".jar" (toString path)))
+      (readTree repository));
+  classpath = lib.concatStringsSep ":" jars;
+in stdenv.mkDerivation rec {
+  pname = "maven-demo";
+  version = "1.0";
+
+  src = builtins.fetchTarball
+    "https://github.com/fzakaria/nixos-maven-example/archive/main.tar.gz";
+  buildInputs = [ maven makeWrapper ];
+
+  buildPhase = ''
+    echo "Using repository ${repository}"
+    mvn --offline -Dmaven.repo.local=${repository} package;
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin
+    mkdir -p $out/share/java
+
+    install -Dm644 target/${pname}-${version}.jar $out/share/java
+    # create a wrapper that will automatically set the classpath
+    # this should be the paths from the dependency derivation
+    makeWrapper ${jre}/bin/java $out/bin/${pname} \
+          --add-flags "-classpath $out/share/java/${pname}-${version}.jar:${classpath}" \
+          --add-flags "Main"
+  '';
+}
+```
+
+### MANIFEST file via Maven Plugin
```

Comment:
Is this a necessity to the above? It should be made clear in general, whether sections are targeted towards developers of a `shell.nix` with maven, or casual contributors to Nixpkgs that just want to add a Java program to Nixpkgs, and compile it from source. Hence, I don't understand in what case one should need to "augment" the `pom.xml` file.

Reply:
I added such sections.